### PR TITLE
Improvise NSGA2

### PR DESCRIPTION
**Pull request by jonpsy**
Continuation of #262 .

The goal of this PR is to increase the efficiency of this code, enhance its expressibility and reducing its code size as much as possible.

Summary of my goals:

* Minimize usage of raw ```for``` loops
* Employing std & arma algorithms whenever possible
* Potential error fixes.
* Introduce OMP parallelism [https://ieeexplore.ieee.org/document/6968779] whether on this class or a separate class.

I have a few questions before I do that, let me know your thoughts on that.

Providing logs for the convenience of the reader. Stay tuned for more progress updates!

**LOGS**

**Main Loop**
- [x] Initial population should be within (lowerBound, upperBound) so that they exist inside the solution space [link](https://github.com/mlpack/ensmallen/pull/263#discussion_r588002920)
- [x] Incorporate CheckArbitraryFunctionTypeAPI<> for NSGA2.
- [x]  Add RequireDenseFloatingType for NSGA2

**Fast Non Dominated Sorting**
- [x] Discard the empty vector appended to the front. [link](https://github.com/mlpack/ensmallen/pull/263#discussion_r587526998)
- [x]  ```fronts.size() > 0``` ==> ```!fronts.empty()```

**Crowding Distance Assignment**
- [x] Pull the ```crowdingDistance``` 0 initialisation out of the function 
- [x] Remove ```fronts.size() > 0 ``` check, since empty front isn't passed at all. Refer the above section.
- [x]  ```crowdingDistance``` is filled using objective values not by taking average of ```crowdingDistance```
- [x]   ~The calculation seems to be very roundabout, need to confirm with @say4n. If you're reading this, can you check this [patch](https://drive.google.com/file/d/1KjrbkkKL6eviF7dUEjsCNr8BCW-6c4rg/view?usp=sharing)~ As discussed, going with the patch.

**Testing**
- [x] Check for different types of matrices, for ex: ```fmat```

### CODE REVIEW

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (jonpsy) [inline include/ensmallen_bits/nsga2/nsga2_impl.hpp:120]: Shouldn't we check if they're inside the upperBound && lowerBound? Am I missing something? 

An example https://github.com/baopng/NSGA-II/blob/29a6ec33f87b32a7fb2596091e1a51c897106e7b/nsga2/problem.py#L20

Comment (jonpsy) [inline include/ensmallen_bits/nsga2/nsga2_impl.hpp:120]: Shouldn't we check if they're inside the upperBound && lowerBound? Am I missing something? 

An example https://github.com/baopng/NSGA-II/blob/29a6ec33f87b32a7fb2596091e1a51c897106e7b/nsga2/problem.py#L20

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (jonpsy) [inline include/ensmallen_bits/nsga2/nsga2_impl.hpp:413]: Setting the same value to the crowdingDistance value of a front multiple times.

**Comment by jonpsy** [comment]
Requesting reviews @rcurtin  @zoq  @ShikharJ since you've mentored the original PR. If I've gone ahead of myself with these changes, please do tell me :).

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (jonpsy) [inline include/ensmallen_bits/nsga2/nsga2_impl.hpp:369]: Caught a big fish here. Allow me to explain myself,

- The while loop breaks when the newFront is empty. Note that the newFront is pushed **regardless** of any of the conditions.
- At the end of our while loop, the nextFront is not updated at all! That means its an empty vector which gets pushed into our ```front```
-  This causes the while loop to terminate

Result?

You got an additional empty vector in your ```front``` set. Which causes smelly code.

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (jonpsy) [inline include/ensmallen_bits/nsga2/nsga2_impl.hpp:401]: Continuing on the "big fish", this causes this redundant if clause because of that empty vector.

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (jonpsy) [inline include/ensmallen_bits/nsga2/nsga2_impl.hpp:402]: Since we're looping over all the fronts, can we not simply initialize ```crowdDistance``` ahead of calling this function?

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (jonpsy) [reply include/ensmallen_bits/nsga2/nsga2_impl.hpp:369]: https://github.com/haris989/NSGA-II/blob/7680fd70c5202d0673aa26b9fb50a8e9c387623b/NSGA%20II.py#L71

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (jonpsy) [reply include/ensmallen_bits/nsga2/nsga2_impl.hpp:401]: https://github.com/haris989/NSGA-II/blob/7680fd70c5202d0673aa26b9fb50a8e9c387623b/NSGA%20II.py#L76, notice there are no checks

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (zoq) [inline include/ensmallen_bits/nsga2/nsga2_impl.hpp:95]: The size of the objectives depends on the size of the population, which can increase in case of a crossover, so I don't think you can get away with a const matrix.

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (zoq) [reply include/ensmallen_bits/nsga2/nsga2_impl.hpp:120]: I have to check the paper, not sure if the constrained was only for the mutation/crossover step or for the initial population as well. Also might be good to reference the paper instead of another python implementation.

**Comment by zoq** [comment]
> Summary of my goals:
> 
> * Minimize usage of raw `for` loops
> * Employing std & arma algorithms whenever possible
> * Potential error fixes.
> * Introduce OMP parallelism [https://ieeexplore.ieee.org/document/6968779] whether on this class or a separate class.

Sounds good to me, if we can show that an OpenMP solution is faster let's integrate it into the current method, but maybe armadillo build against openblas is already faster, haven't read the paper yet.

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (jonpsy) [reply include/ensmallen_bits/nsga2/nsga2_impl.hpp:120]:
```c++
	  
	  /*if limits are specified it generates the value in 
	    range of minimum and maximum value of the variable*/
	  else
	    {
	      pop_ptr->ind[i].xreal[j] = d*(lim_r[j][1] - lim_r[j][0])+lim_r[j][0];
	    }
```

Source: IITK Kanpur Genetics Laboratory, Original Implementation in C

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (jonpsy) [reply include/ensmallen_bits/nsga2/nsga2_impl.hpp]: The sorting part inside the CrowdingDistance method not implemented.

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (conradsnicta) [inline include/ensmallen_bits/nsga2/nsga2_impl.hpp:146]: I'd be wary of replacing simple loops with algorithms from the std namespace. I don't think there is any performance advantage -- in the end the code will do exactly the same thing.  At the same time, this change makes things harder to read/understand, which adds to the technical debt (ie. increases the maintenance burden).

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (conradsnicta) [inline include/ensmallen_bits/nsga2/nsga2_impl.hpp:156]: Again, a simple and effective 'for' loop was replaced with something more elaborate.  There is no need for this.

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (jonpsy) [reply include/ensmallen_bits/nsga2/nsga2_impl.hpp:146]: I had a feeling I was overdoing it. Thanks for the review.

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (jonpsy) [reply include/ensmallen_bits/nsga2/nsga2_impl.hpp:156]: Are you talking about the std::fill or std::for_each?

Comment (jonpsy) [reply include/ensmallen_bits/nsga2/nsga2_impl.hpp:156]: Because I moved crowdingDistance zero intialisation out from the function. So instead of setting 0 to that vector per front, I did it at once. 

If however you're talking about std::for_each, I'll revert that change.

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (zoq) [reply include/ensmallen_bits/nsga2/nsga2_impl.hpp:381]: Nice catch!

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (zoq) [inline include/ensmallen_bits/nsga2/nsga2_impl.hpp:130]:
```suggestion
    // Ensure population is within variable space.
    for (size_t geneIdx = 0; geneIdx < numVariables; geneIdx++)
    {
      if (population[i][geneIdx] < lowerBound[geneIdx])
        population[i][geneIdx] = lowerBound[geneIdx];
      
      if (population[i][geneIdx] > upperBound[geneIdx])
        population[i][geneIdx] = upperBound[geneIdx];
    }
```

We tend to be picky about the style.

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (zoq) [inline include/ensmallen_bits/nsga2/nsga2_impl.hpp:190]:
```suggestion
    // Discards unfit population from the R_{t} to yield P_{t+1}.
```

Pedantic style fix.

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (zoq) [inline include/ensmallen_bits/nsga2/nsga2_impl.hpp:380]:
```suggestion
  // Remove the empty final set.
```

Another minor style fix, not sure I mentioned this at some point, but ensmallen follows https://github.com/mlpack/mlpack/wiki/DesignGuidelines.

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (zoq) [inline include/ensmallen_bits/nsga2/nsga2_impl.hpp:164]: I guess before the reset to zero, we got lucky, because in the `CrowdingDistanceAssignment` method we use `front[i + 1]` which should be undefined:

```
crowdingDistance[front[i]] += (crowdingDistance[front[i - 1]] -
            crowdingDistance[front[i + 1]])
```

Comment (zoq) [reply include/ensmallen_bits/nsga2/nsga2_impl.hpp:164]: Ahh, nevermind we set it to zero before as well.

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (zoq) [inline include/ensmallen_bits/nsga2/nsga2_impl.hpp:381]: Looks like this causes a bad alloc.

Comment (jonpsy) [reply include/ensmallen_bits/nsga2/nsga2_impl.hpp:381]: The bad alloc is because I used ```!fronts.empty()``` instead of ```!fronts[i].empty()```. Lemme fix that ;)

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (jonpsy) [reply include/ensmallen_bits/nsga2/nsga2_impl.hpp:164]: @zoq  Regarding this, I think this implementation is wrong. Here's an excerpt from the paper

crowdingdistance

Notice the equation should've been 
```c++
crowdingDistance[front[i]] += (fValue[i+1] - fValue[i-1])/(maxFval - minFval)
```

and not the average of crowdingDistances

Comment (jonpsy) [reply include/ensmallen_bits/nsga2/nsga2_impl.hpp:164]: I've implemented it in this  [patch](https://drive.google.com/file/d/1KjrbkkKL6eviF7dUEjsCNr8BCW-6c4rg/view)

Comment (zoq) [reply include/ensmallen_bits/nsga2/nsga2_impl.hpp:164]: Good, can you make the patch part of the PR?

Comment (jonpsy) [reply include/ensmallen_bits/nsga2/nsga2_impl.hpp:164]: I've no problem with that. Thing is, @say4n  has implemented assignment first and sorting later which I think is wrong but I guess he must have some reason. I was waiting for his review, very well, I'll add this patch on my PR.

Comment (say4n) [reply include/ensmallen_bits/nsga2/nsga2_impl.hpp:164]: I think I had unrolled the first iteration of some parts of the optimisation loop to be outside the for loop. This had the benefit of instantiating some items as far as I recall. I don't remember all the details now so do take this with a grain of salt! :)

Comment (jonpsy) [reply include/ensmallen_bits/nsga2/nsga2_impl.hpp:164]: Thanks for clarifying :). I'd say let's merge my patch for now as it is straightforward & adheres to the research paper directly. We can then try your optimization method and see how it all fits in.

Comment (say4n) [reply include/ensmallen_bits/nsga2/nsga2_impl.hpp:164]: Yes, for sure! Great work btw!!

**Comment by jonpsy** [comment]
All the tests pass, Wohoo!. All the tasks are done as well. @zoq Do you have any comments?

**Comment by zoq** [comment]
> All the tests pass, Wohoo!. All the tasks are done as well. @zoq Do you have any comments?

I'll do another pass over the PR, in the meantime can you update the `HISTORY.md` so it passes the Travis CI check.

File: include/ensmallen_bits/nsga2/nsga2.hpp
Comment (zoq) [inline include/ensmallen_bits/nsga2/nsga2.hpp:183]: I'm not sure I see the reason to change the name of the parameter here, but I don't mind to change it, but let's choose a more descriptive name here, I mean `population` is clear, but `Fval` is not really a parameter name that describes what the parameter is about. Let me know what you think.

File: include/ensmallen_bits/nsga2/nsga2.hpp
Comment (zoq) [inline include/ensmallen_bits/nsga2/nsga2.hpp:253]: See comment above.

File: include/ensmallen_bits/nsga2/nsga2.hpp
Comment (zoq) [inline include/ensmallen_bits/nsga2/nsga2.hpp:269]: See comment above.

File: include/ensmallen_bits/nsga2/nsga2.hpp
Comment (zoq) [inline include/ensmallen_bits/nsga2/nsga2.hpp:289]: See comment above.

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (zoq) [inline include/ensmallen_bits/nsga2/nsga2_impl.hpp:168]: Looks like the indentation is wrong here, it should be two spaces, and the next line should tabbed twice (4 spaces).

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (zoq) [inline include/ensmallen_bits/nsga2/nsga2_impl.hpp:409]: Should be tabbed twice (4 spaces).

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (zoq) [inline include/ensmallen_bits/nsga2/nsga2_impl.hpp:405]: Should we use `fValues` here?

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (zoq) [inline include/ensmallen_bits/nsga2/nsga2_impl.hpp:416]: See comments above.

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (zoq) [inline include/ensmallen_bits/nsga2/nsga2_impl.hpp:420]: Should be tabbed twice (4 spaces).

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (zoq) [inline include/ensmallen_bits/nsga2/nsga2_impl.hpp:429]: Can you check the indentation here?

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (zoq) [inline include/ensmallen_bits/nsga2/nsga2_impl.hpp:404]: Looks like it would be possible to use an `arma::vec` here, unless I missed something and doing so would make sorting more complex?

Comment (jonpsy) [reply include/ensmallen_bits/nsga2/nsga2_impl.hpp:404]: Nice catch! :)

File: include/ensmallen_bits/nsga2/nsga2.hpp
Comment (jonpsy) [reply include/ensmallen_bits/nsga2/nsga2.hpp:183]: My reasoning is ```calculatedObjectives``` is too vague and lengthy on top of it. 

_"calculatedObjectives of what?"_  A user might ask.

Perhaps ```Fval``` is bad naming, how about ```populationFitness```?

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (jonpsy) [reply include/ensmallen_bits/nsga2/nsga2_impl.hpp:405]: Indeed, we need to store the fValues for the current objective, or are you talking about the naming?

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (jonpsy) [reply include/ensmallen_bits/nsga2/nsga2_impl.hpp:420]: So, if we're doing an assignment in a new line we should tab twice, right?

**Comment by jonpsy** [comment]
Perhaps a good idea to add this too?

```c++
  traits::CheckArbitraryFunctionTypeAPI<
      FunctionType, BaseMatType>();
  RequireDenseFloatingPointType<BaseMatType>();
```

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (zoq) [reply include/ensmallen_bits/nsga2/nsga2_impl.hpp]: I was talking about the name so the comment matches the parameter name.

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (zoq) [reply include/ensmallen_bits/nsga2/nsga2_impl.hpp]: Yes, everything that doesn't fit in the 80 character limit has to be tabbed twice.

**Comment by jonpsy** [comment]
AppVeyor is failing, but the tests work, can you throw some light on this please?

**Comment by zoq** [comment]
Strange, I have to debug this on a local machine, but looks like the windows build never passed on this PR.

**Comment by jonpsy** [comment]
On second thought, I think its better to keep ```calculatedObjectives``` because fitness doesn't always equate to calculated objective value. For ex: SPEA-II has a clearly distinguished fitness and objective value.

**Comment by jonpsy** [comment]
Let's have #275 merged, so that all the test pass. Had to force-push because of SchafferN1 fix (another force-push coming for because of fonseca =P). Cheers!

**Comment by jonpsy** [comment]
The double and float comparison is causing some really interesting test failures.

**Review by zoq — COMMENTED** [review]
Looks like you changed the appveyor build settings from debug to release, does that mean ensmallen doesn't build in debug mode?

**Comment by jonpsy** [comment]
No, I was just messing around(for appveyor RDP to Windows) . It's really weird that it fails on SchafferN1Double but passes on SchafferN1DoubleVector!? 

Similarly, passes for SchafferN1Float but fails on SchafferN1FloatVector.

**Comment by jonpsy** [comment]
The ```numGeneration``` value is suspiciously high in our tests (5000), but this confirms it. The paper (and any MOEAD algorithm for that matter) uses ~250 ```numGenerations```. What happened was, this high number of generation lead to a "overfit", our pareto points totally lost diversity.

Infact, there were many coinciding points which is a nightmare for any MOO algorithm. I fixed the default arguments in the test for ```mutationRate``` , ```crossoverProb``` and put the ```numGenerations``` as 250 (see figure below). All the tests pass with flying colors on VS2019.

image

**Comment by jonpsy** [comment]
Test seems to pass when I run them manually on Appveyor but fails when run through CTest (MSVC-2015,17). Tests pass in both CTest and manually on MSVC19 and Linux builds.

testpassed

**Comment by zoq** [comment]
> Test seems to pass when I run them manually on Appveyor but fails when run through CTest (MSVC-2015,17). Tests pass in both CTest and manually on MSVC19 and Linux builds. @zoq, is there something different in CTest that I might be missing?
> 
> CroppedNSGA2Testpass2015

Nothing, no, this could just be a bad seed, did you run the tests locally like 1000x with a different seed to check if the test pass each time? Note you have to uncomment:

https://github.com/mlpack/ensmallen/blob/9f90ec92aa64fa03b83e1f5206af198d73f4359c/tests/main.cpp#L25-L27

**Comment by zoq** [comment]
Hm, not sure I can see the fix in the last commits, was this just a strange windows fluke?

**Comment by jonpsy** [comment]
I changed the tolerance from 1e-3 to 0.1, the problem was there were was one outlier in the Pareto front ```-0.025...``` which failed the test every time. (Bound is [0, 2])

Reg it being a fluke, well, it must be one really lucky fluke to be not caught in [3000](https://github.com/mlpack/ensmallen/pull/263/commits/f3f699a50879d56156082c98a7a2c48f9a4fd3ed) (this change passed the test) attempts :). I tried this stress testing on my local MSVC17,19 as well. 

Notice that the ```numGeneration``` parameter is now 300 because NSGA2 performance caps around this, no need of 4000.

Just to be 100% sure, I will share the pastebin for stress testing inside the Appveyor.

**Comment by jonpsy** [comment]
> Hm, not sure I can see the fix in the last commits, was this just a strange windows fluke?

Your question is well-founded, it looks like Appveyor cmd does not repeat for 3000 times, even after putting ```---repeat-until-fail``` command. So I went ahead and run it manually for like ~50 times. Here's the pastebin (https://pastebin.com/k641DjC7). Let me know if you require further proof.

**Comment by zoq** [comment]
> > Hm, not sure I can see the fix in the last commits, was this just a strange windows fluke?
> 
> Your question is well-founded, it looks like Appveyor cmd does not repeat for 3000 times, even after putting `---repeat-until-fail` command. So I went ahead and run it manually for like ~50 times for MSVC15. Here's the pastebin (https://pastebin.com/k641DjC7). For 100 times: https://pastebin.com/t4yMGdec.
> 
> Let me know if you require further proof.

Thanks, I just was just wondering what it was, thanks for tracking this down.

**Comment by zoq** [comment]
I restarted the windows build. because the build failed, not sure it's because of the RDP session.

**Comment by jonpsy** [comment]
The build fails because of RDP. Its waiting for us to access it, and terminates when we dont. @zoq , do we agree the algorithm is working? If so, I'll revert RDP changes and the build would pass.

**Comment by zoq** [comment]
> The build fails because of RDP. Its waiting for us to access it, and terminates when we dont. @zoq , do we agree the algorithm is working? If so, I'll revert RDP changes and the build would pass.

Ahh, right I have seen you removed it from one section, but looks like not from the other one. This looks good, so let's remove it.

**Comment by jonpsy** [comment]
Hi @zoq , is this GTM?

**Review by zoq — COMMENTED** [review]
This looks really good to me, just one more really minor style issue.

File: include/ensmallen_bits/nsga2/nsga2_impl.hpp
Comment (zoq) [inline include/ensmallen_bits/nsga2/nsga2_impl.hpp:176]: Can you use the same indentation here you used for the other `sort` and `transform`.

**Comment by jonpsy** [comment]
Once #283 is merged, I will incorporate that feature along with the rebase changes + static analysis fix changes.

**Comment by zoq** [comment]
> Once #283 is merged, I will incorporate that feature along with the rebase changes + static analysis fix changes.

Nice, just merged #283.

**Comment by jonpsy** [comment]
All right, waiting for the green tick :) .

**Comment by jonpsy** [comment]
All green! 🥳

**Review by zoq — APPROVED** [review]
Thanks for another great contribution :+1:

**Comment by zoq** [comment]
Thanks for digging into the optimization method, really nice improvements.

**Pull request merged**
